community/rfcs/techniques/design/20200306-weight-clustering-design.md

```diff
@@ -0,0 +1,270 @@
+}
+
+# Clustering a sequential model
+model = keras.Sequential([
+    layers.Dense(10, input_shape=(10,)),
+    layers.Dense(10),
+])
+
+# Train model here
+# ...
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+clustered_model.build(input_shape=(1, 10))
+```
+
+```python
+# Clustering a functional model
+i1 = keras.Input(shape=(10,))
+i2 = keras.Input(shape=(10,))
+x1 = layers.Dense(10)(i1)
+x2 = layers.Dense(10)(i2)
+outputs = layers.Add()([x1, x2])
+
+model = keras.Model(inputs=[i1, i2], outputs=outputs)
+
+# Train model here
+# ...
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+```
+
+#### Cluster Some Layers
+
+Individual layers within a model can be clustered selectively by invoking `cluster_weights()` on a layer object, as follows:
+
+```python
+i1 = keras.Input(shape=(10,))
+i2 = keras.Input(shape=(10,))
+x1 = cluster.cluster_weights(keras.layers.Dense(10), **clustering_params)(i1)
+x2 = keras.layers.Dense(10)(i2)
+outputs = keras.layers.Add()([x1, x2])
+
+clustered_model = keras.Model(inputs=[i1, i2], outputs=outputs)
+```
+In the above example model the weights of the 1st `Dense` layer will be clustered, whereas the weights of the 2nd layer will not be clustered.
+
+NOTE: Since clustering can only be applied to a pre-trained model, in real-life selective clustering use cases, weights will have to be loaded into the model at creation.
+
+#### Export Clustered Model
+
+```python
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+stripped_model = cluster.strip_clustering(clustered_model)
+keras.models.save_model(stripped_model, keras_file)
+```
+
+#### Import Clustered Model
+
+A clustered Keras model needs to be loaded within a `cluster_scope` in order to be successfully deserialized:
+
+```python
+with cluster_scope():
+  loaded_model = keras.models.load_model(keras_file)
+```
+
+## Compatibilty
+
+### At Launch
+
+Users can apply weight clustering with the following APIs:
+
+* Model building: `tf.keras` with only Sequential and Functional models
+* TensorFlow versions: TF 1.x for versions 1.14+ and 2.x.
+* TensorFlow execution mode: both graph and eager
+
+### Roadmap
+
+This is a list of possible future improvements and features that are currently under consideration:
+
+* Support for custom clusterable layers &mdash; implemented already to some extent, but the API needs further refinement.
+* Sparsity-aware clustering, to ensure that the sparsity of pruned models is not destroyed during the clustering and fine-tuning process.
+* Clustering schedule and callback, to provide more advanced training-time control of the clustering behaviour and progressive layer clustering throughout training.
+* Improved training / fine-tuning by updating cluster-to-weight associations during the fine-tuning phase. This one is likely to be a good use case for having support for clustering schedule in the API.
+* More cluster centroid initialization methods, e.g. k-means.
```

Comment:
This is a typo. Should be k-means ++, right?

Reply:
K-means ++ is a good thing to try. Also, it is worth mentioning that we will work on the improvement of density-based initialization methods because this differentiates clustering from QAT the most.